src/libstd/sys/wasi/fs.rs

```diff
@@ -1,294 +1,689 @@
+    // This seems to not be in wasi's API yet, and we may need to end up
+    // emulating it ourselves. For now just return an error.
     unsupported()
 }
 
-pub fn copy(_from: &Path, _to: &Path) -> io::Result<u64> {
-    unsupported()
+fn open_at(fd: &WasiFd, path: &Path, opts: &OpenOptions) -> io::Result<File> {
+    let fd = fd.open(
+        opts.dirflags,
+        path.as_os_str().as_bytes(),
+        opts.oflags,
+        opts.rights_base(),
+        opts.rights_inheriting(),
+        opts.fdflags,
+    )?;
+    Ok(File { fd })
+}
+
+// FIXME: we shouldn't implement this. It'd be much better to share this between
+// libc (the wasi-sysroot) and Rust as the logic here is likely far more tricky
+// than what we're executing below. For now this is a stopgap to enable this
+// module, but we should add an official API in upstream wasi-libc which looks
+// like this.
+//
+// In the meantime this is highly unlikely to be correct. It allows some basic
+// testing but is not at all robust.
+fn open_parent(p: &Path) -> io::Result<(&'static WasiFd, &Path)> {
+    let map = preopened_map();
+    for ancestor in p.ancestors() {
+        if let Some(fd) = map.get(ancestor) {
+            let tail = p.strip_prefix(ancestor).unwrap();
+            let tail = if tail == Path::new("") {
+                ".".as_ref()
+            } else {
+                tail
+            };
+            return Ok((fd, tail))
+        }
+    }
+    let msg = format!("failed to find a preopened file descriptor to open {:?}", p);
+    return Err(io::Error::new(io::ErrorKind::Other, msg));
+
+    type Preopened = HashMap<PathBuf, ManuallyDrop<WasiFd>>;
+    fn preopened_map() -> &'static Preopened {
+        static PTR: AtomicPtr<Preopened> = AtomicPtr::new(ptr::null_mut());
+        unsafe {
+            let ptr = PTR.load(SeqCst);
+            if !ptr.is_null() {
+                return &*ptr;
+            }
+
+            let mut map = Box::new(HashMap::new());
+            for fd in 3.. {
+                let mut buf = mem::zeroed();
+                if cvt_wasi(libc::__wasi_fd_prestat_get(fd, &mut buf)).is_err() {
+                    break;
+                }
+                if buf.pr_type != libc::__WASI_PREOPENTYPE_DIR {
+                    continue;
+                }
+                let len = buf.u.dir.pr_name_len;
+                let mut v = vec![0u8; len];
+                let res = cvt_wasi(libc::__wasi_fd_prestat_dir_name(
+                    fd,
+                    v.as_mut_ptr() as *mut i8,
+                    v.len(),
+                ));
+                if res.is_err() {
+                    continue;
+                }
+                let path = PathBuf::from(OsString::from_vec(v));
+                map.insert(path, ManuallyDrop::new(WasiFd::from_raw(fd)));
+            }
+            let ptr = Box::into_raw(map);
+            match PTR.compare_exchange(ptr::null_mut(), ptr, SeqCst, SeqCst) {
```

Comment:
A comment here explaining how if we lost the race to initialize, we drop our local map would be good. I feel  like I always have to double check for this kind of thing and atomics can be so tricky...

Reply:
But yes -- getting this shared with upstream would be A+++!